### PR TITLE
Maximum weighted matching

### PR DESCRIPTION
**Pull request by mburq**
Hi,

This is a formulation for maximum weighted matching, works for both bipartite and non-bipartite graphs. 
In the case of bipartite graphs, the solution of the LP is automatically integral (i.e. no need to enforce integrality constraints).
In the case of non-bipartite graphs, it requires an MIP solver (e.g. GLPK) and may therefore be much slower to solve to optimality.

Disclaimer: This is the first time I do a pull request, so I hope I'm not missing an important step.

### CODE REVIEW

File: src/matching/matching.jl [hunk 1]
```diff
@@ -1,7 +1,7 @@
 __precompile__(true)
 module Matching
 using LightGraphs
-export MatchingResult, maximum_weight_maximal_matching, minimum_weight_perfect_matching
+export MatchingResult, max_weight_matching, maximum_weight_maximal_matching, minimum_weight_perfect_matching
```
Comment (CarloLucibello) [inline src/matching/matching.jl]: max_weight_matching->maximum_weight_matching

for coerence with the other functions

File: src/matching/max_weight_matching.jl [hunk 2]
```diff
@@ -0,0 +1,59 @@
+"""
+max_weight_matching{T <:Real}(g::Graph, w::Dict{Edge,T} = Dict{Edge,Int64}())
+
+Given a graph `g` and an edgemap `w` containing weights associated to edges,
+returns a matching with the maximum total weight.
+`w` is a dictionnary that maps edges i => j (with i <= j) to weights.
```
Comment (CarloLucibello) [inline src/matching/max_weight_matching.jl]: typo here

**Comment by CarloLucibello** [comment]
<- If the graph is not bipartite, then the computation time may grow exponentially.

I do not see integer programming part in case of fractionary solution, why do you say that? 

I see three options here:
- @assert is_bipartite(g)
- accept also bipartite graphs, but caution the user about the possible suboptimality of the solution, possibly returning also a fractional_solution=true flag
- add an Integer programming part in case of fractional solutions

**Comment by CarloLucibello** [comment]
also, coverage should be increased

**Comment by mburq** [comment]
Sorry I wasn't clear in my comments, the integer programming is taken care of by the solver (by using the _Int_ keyword in @variable, line 45).  
Good idea to use is_bipartite(g) so that it doesn't require a an Integer Programming solver if the graph is bipartite.

File: src/matching/maximum_weight_matching.jl [hunk 3]
```diff
@@ -0,0 +1,71 @@
+"""
+maximum_weight_matching{T <:Real}(g::Graph, w::Dict{Edge,T} = Dict{Edge,Int64}())
+
+Given a graph `g` and an edgemap `w` containing weights associated to edges,
+returns a matching with the maximum total weight.
+`w` is a dictionnary that maps edges i => j to weights.
```
Comment (CarloLucibello) [inline src/matching/maximum_weight_matching.jl]: dictionnary->dictionary

File: src/matching/maximum_weight_matching.jl [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+The efficiency of the algorithm depends on the input graph:
+  - If the graph is bipartite, then the LP relaxation is integral.
+  - If the graph is not bipartite, then it requires a MIP solver and
+  the computation time may grow exponentially.
+
+The package JuMP.jl and one of its supported solvers is required.
+
+Returns MatchingResult containing:
+  - a solve status (indicating whether the problem was solved to optimality)
```
Comment (CarloLucibello) [inline src/matching/maximum_weight_matching.jl]: MatchingResult doesn't contain solve

File: src/matching/maximum_weight_matching.jl [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  - the optimal cost
+  - a list of each node's match.
```
Comment (CarloLucibello) [inline src/matching/maximum_weight_matching.jl]: -> a list of each vertex's match (or -1 for unmatched vertices)

File: src/matching/maximum_weight_matching.jl (hunk 4 above)
Comment (CarloLucibello) [inline src/matching/maximum_weight_matching.jl]: ok, I saw the TODO comment

File: test/matching/runtests.jl [hunk 6]
```diff
@@ -56,6 +56,57 @@ match = Matching.maximum_weight_maximal_matching(g,w,0)
...
+
+g = Graph(4)
+w =Dict{Edge,Float64}()
+add_edge!(g, 1,3)
+add_edge!(g, 1,4)
+add_edge!(g, 2,4)
+w[Edge(1,3)] = 1
+w[Edge(1,4)] = 3
+w[Edge(2,4)] = 1
+match = Matching.maximum_weight_matching(g,w)
```
Comment (CarloLucibello) [inline test/matching/runtests.jl]: can you remove Matching. from here and from the other lines in this file (also a check to see that everything is correctly exported)

**Comment by CarloLucibello** [comment]
>Sorry I wasn't clear in my comments, the integer programming is taken care of by the solver (by using the Int keyword in @variable, line 45).
Good idea to use is_bipartite(g) so that it doesn't require a an Integer Programming solver if the graph is bipartite.

Ok, thanks for the clarification, my knowledge of JuMP is very limited. This looks good, i can be merged once the remaining comments are addressed

**Comment by CarloLucibello** [comment]
this case should be tested
```julia
if setdiff(edge_list, keys(w)) != [] # If some edges do not have a key in w.
      error("Some edge weights are missing, check that keys i => j in w satisfy i <= j")
end
```
with a
```julia
@test_throws ErrorException ....
```

**Comment by mburq** [comment]
Just a quick follow-up: can this PR be merged?
Thanks @CarloLucibello for the help in the review process.

**Pull request merged**